packages/bun-vscode/src/features/tests/bun-test-controller.ts

```diff
@@ -0,0 +1,1539 @@
+    } else {
+      bunCommand = cmd;
+    }
+
+    const testArgs = args.length ? args : ["test"];
+    if (customFlag) {
+      testArgs.push(customFlag);
+    }
+
+    return { bunCommand, testArgs };
+  }
+
+  private async staticDiscoverTests(testItem?: vscode.TestItem | false, filePath?: string): Promise<void> {
+    let targetPath = filePath;
+    if (!targetPath && testItem) {
+      targetPath = testItem?.uri?.fsPath || this.workspaceFolder.uri.fsPath;
+    }
+    if (!targetPath) {
+      return;
+    }
+
+    try {
+      const fileContent = await fs.readFile(targetPath, "utf8");
+      const testNodes = this.parseTestBlocks(fileContent);
+
+      const fileUri = vscode.Uri.file(windowsVscodeUri(targetPath));
+      let fileTestItem = testItem || this.testController.items.get(windowsVscodeUri(targetPath));
+      if (!fileTestItem) {
+        fileTestItem = this.testController.createTestItem(
+          fileUri.toString(),
+          path.relative(this.workspaceFolder.uri.fsPath, targetPath),
+          fileUri,
+        );
+        this.testController.items.add(fileTestItem);
+      }
+      fileTestItem.children.replace([]);
+
+      this.addTestNodes(testNodes, fileTestItem, targetPath);
+    } catch (err) {
+      output.appendLine(`Error reading test file at ${targetPath}: ${err}`);
+    }
+  }
+
+  private parseTestBlocks(fileContent: string): TestNode[] {
+    const cleanContent = fileContent
+      .replace(/\/\*[\s\S]*?\*\//g, match => match.replace(/[^\n\r]/g, " "))
+      .replace(/\/\/.*$/gm, match => " ".repeat(match.length));
+
+    const testRegex =
+      /\b(describe|test|it)(?:\.(?:skip|todo|failing|only))?(?:\.(?:if|todoIf|skipIf)\s*\([^)]*\))?(?:\.each\s*\([^)]*\))?\s*\(\s*(['"`])((?:\\\2|.)*?)\2\s*,/g;
+
+    const stack: TestNode[] = [];
+    const root: TestNode[] = [];
+    let match: RegExpExecArray | null;
+
+    match = testRegex.exec(cleanContent);
+    while (match !== null) {
+      const [full, type, , name] = match;
+      const line = cleanContent.slice(0, match.index).split("\n").length - 1;
+
+      while (
+        stack.length > 0 &&
+        match.index > stack[stack.length - 1].startIdx &&
+        this.getBraceDepth(cleanContent, stack[stack.length - 1].startIdx, match.index) <= 0
+      ) {
+        stack.pop();
+      }
+
+      const expandedNodes = this.expandEachTests(full, name, cleanContent, match.index, type as TestNode["type"], line);
+
+      for (const node of expandedNodes) {
+        if (stack.length === 0) {
+          root.push(node);
+        } else {
+          stack[stack.length - 1].children.push(node);
+        }
+
+        if (type === "describe") {
+          stack.push(node);
+        }
+      }
+      match = testRegex.exec(cleanContent);
+    }
+
+    return root;
+  }
+
+  private getBraceDepth(content: string, start: number, end: number): number {
+    const section = content.slice(start, end);
+    const openBraces = (section.match(/\{/g) || []).length;
```

Comment:
this does not work with escape sequences, for example inside of template strings.